### PR TITLE
Kernel: Correctly decode proc_file_type from identifier

### PR DESCRIPTION
**Pull request by timschumi**
inode identifiers in ProcFS are encoded in a way that the parent ID is shifted 12 bits to the left and the PID is shifted by 16 bits.
This means that the rightmost 12 bits are reserved for the file type or the fd.

Since the to_fd and to_proc_file_type decoders only decoded the rightmost 8 bits, decoded values would wrap around
beyond values of 255, resulting in a different value compared to what was originally encoded.

Since fds are offset by +40 when encoded, this caused wrong behaviour when interacting with fds that have a number of 216 or higher.

Fixes: #7820

### CODE REVIEW

**Comment by ambroff** [comment]
I'm new to the project so I don't know what the norm is for this, but you could easily add a regression test for this like so:

```diff
diff --git a/Tests/Kernel/TestProcFS.cpp b/Tests/Kernel/TestProcFS.cpp
new file mode 100644
index 000000000000..5562fea42dda
--- /dev/null
+++ b/Tests/Kernel/TestProcFS.cpp
@@ -0,0 +1,34 @@
+#include <LibTest/TestCase.h>
+#include <cstring>
+#include <unistd.h>
+
+TEST_CASE(test_process_fd_readlink)
+{
+    char buf[MAXPATHLEN];
+
+    // Read the symlink value for stdin, stdout and stderr
+    auto link_length = readlink("/proc/self/fd/0", buf, sizeof(buf));
+    buf[link_length] = '\0';
+    EXPECT_EQ(link_length, 10);
+    EXPECT_EQ(0, strcmp(buf, "/dev/pts/0"));
+
+    link_length = readlink("/proc/self/fd/1", buf, sizeof(buf));
+    buf[link_length] = '\0';
+    EXPECT_EQ(link_length, 10);
+    EXPECT_EQ(0, strcmp(buf, "/dev/pts/0"));
+
+    link_length = readlink("/proc/self/fd/2", buf, sizeof(buf));
+    buf[link_length] = '\0';
+    EXPECT_EQ(link_length, 10);
+    EXPECT_EQ(0, strcmp(buf, "/dev/pts/0"));
+
+    // Create a new file descriptor that is a dup of 0 with the value 255 in order to reproduce issue #8090.
+    auto new_fd = dup2(0, 255);
+    EXPECT_EQ(new_fd, 255);
+
+    // We should get the same link value for fd 255.
+    link_length = readlink("/proc/self/fd/255", buf, sizeof(buf));
+    buf[link_length] = '\0';
+    EXPECT_EQ(link_length, 10);
+    EXPECT_EQ(0, strcmp(buf, "/dev/pts/0"));
+}
```

**Comment by timschumi** [comment]
> I'm new to the project so I don't know what the norm is for this, but you could easily add a regression test for this like so:
> 
> -- snip --

Cool! Would you mind if I extend it with a few more test cases (since fd 255 being an issue was basically just by chance) and add it to this PR (obviously with your authorship, I guess)?

**Comment by ambroff** [comment]
> Cool! Would you mind if I extend it with a few more test cases (since fd 255 being an issue was basically just by chance) and add it to this PR (obviously with your authorship, I guess)?

Of course go ahead.

**Comment by timschumi** [comment]
Alright, I extended the test with a few more edge-cases and added it to the PR.

@ambroff Can you double-check if the [commit metadata](https://github.com/SerenityOS/serenity/commit/2f98ae387efd42120713440e4ba3e0d6f57ad09d.patch) matches the data that you would usually use (since I went by the public GitHub data to fill the name and E-Mail)?

**Comment by ambroff** [comment]
> 
> 
> Alright, I extended the test with a few more edge-cases and added it to the PR.
> 
> @ambroff Can you double-check if the [commit metadata](https://github.com/SerenityOS/serenity/commit/2f98ae387efd42120713440e4ba3e0d6f57ad09d.patch) matches the data that you would usually use (since I went by the public GitHub data to fill the name and E-Mail)?

That looks fine, thanks for the attribution.

**Comment by ambroff** [comment]
The tests failed because in CI the device used for the controlling terminal isn't `/dev/pts/0` but is something like `device:4,64 (SerialDevice)`.

So you might have to change it to be something like.

1. read link for `/proc/self/fd/0` as `expected_link_value`
2. Compare the link value for all other duped fds with `expected_link_value`.

**Comment by timschumi** [comment]
Suggestion applied, thanks again!

Let's see if the tests are successful now...
